### PR TITLE
Polars: fill_nan

### PR DESCRIPTION
**Pull request by Shoeboxam**

### CODE REVIEW

**Comment by Shoeboxam** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opendp/opendp/1590?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1534** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1534?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1539** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1539?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1540** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1540?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1586** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1586?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1590** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1590?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1584** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1584?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1542** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1542?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1498** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1498?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1479** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1479?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1496](https://github.com/opendp/opendp/pull/1496) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1496?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#908** <a href="https://app.graphite.dev/github/pr/opendp/opendp/908?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#913](https://github.com/opendp/opendp/pull/913) <a href="https://app.graphite.dev/github/pr/opendp/opendp/913?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#890** <a href="https://app.graphite.dev/github/pr/opendp/opendp/890?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#856](https://github.com/opendp/opendp/pull/856) <a href="https://app.graphite.dev/github/pr/opendp/opendp/856?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#829** <a href="https://app.graphite.dev/github/pr/opendp/opendp/829?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#854](https://github.com/opendp/opendp/pull/854) <a href="https://app.graphite.dev/github/pr/opendp/opendp/854?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#888](https://github.com/opendp/opendp/pull/888) <a href="https://app.graphite.dev/github/pr/opendp/opendp/888?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1519** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1519?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1506](https://github.com/opendp/opendp/pull/1506) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1506?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1566** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1566?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by mccalluc — COMMENTED** [review]
Mostly concerned about clarifying null vs. nan.

File: rust/src/transformations/make_stable_expr/expr_fill_nan/mod.rs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            "interior metrics on the input and fill expressions must match: {:?} != {:?}",
+            data_metric,
+            fill_metric
+        );
+    }
+
+    let fill_series = fill_domain.active_series()?;
+    let fill_can_be_nan = match &fill_series.field.dtype {
+        DataType::Float32 => fill_series.atom_domain::<f32>()?.nullable(),
+        DataType::Float64 => fill_series.atom_domain::<f64>()?.nullable(),
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/expr_fill_nan/mod.rs:65]: I think it's still well-named from the perspective of atom domain. Not opening an issue for the time being.

File: rust/src/transformations/make_stable_expr/expr_fill_nan/mod.rs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        _ => return fallible!(MakeTransformation, "fill_nan filler must be float data"),
+    };
+
+    if fill_can_be_nan {
+        return fallible!(MakeTransformation, "fill_nan requires non-nan fill data");
+    }
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/expr_fill_nan/mod.rs:71]: I adjusted the wording to distinguish the two error cases.

File: rust/src/transformations/make_stable_expr/expr_fill_nan/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/expr_fill_nan/mod.rs:65]: Added a comment

File: rust/src/transformations/make_stable_expr/mod.rs [hunk 3]
```diff
@@ -74,6 +77,10 @@ where
         input_domain: ExprDomain,
         input_metric: M,
     ) -> Fallible<Transformation<ExprDomain, ExprDomain, M, M>> {
+        if expr_fill_nan::match_fill_nan(&self).is_some() {
+            return expr_fill_nan::make_expr_fill_nan(input_domain, input_metric, self);
+        }
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/mod.rs:82]: `test_make_expr_fill_nan` passes through this

**Review by mccalluc — APPROVED** [review]
Thanks for responses.

**Pull request merged**